File: modules/dms-vision/android/src/main/java/expo/modules/dmsvision/SelfTest.kt
// The native half of the self-test (README §8.3). It runs the PRODUCTION classes (FeatureExtractor,
// Roi, HeadPose, RecordEncoder, Batcher, GazeInputAssembler, SubjectStatisticTracker, Focal, and the
// build's GazeNetFactory through `makeNet`) over the golden vectors and returns their outputs; JS
// diffs them. It never carries a copy of any production logic. JVM only (org.json is part of Android).

package expo.modules.dmsvision

import java.nio.ByteBuffer
import org.json.JSONArray
import org.json.JSONObject

object SelfTest {
  const val PLATFORM = "android"

  /**
   * `vectorsJson`: a JSON array of vectors. `gazeNetAvailable` / `makeNet` come from the build's
   * GazeNetFactory. Throws DmsError.badArgs if the input is not a JSON array at all.
   */
  fun run(vectorsJson: String, gazeNetAvailable: Boolean, makeNet: () -> GazeNetRunner?): String {
    val vectors = try {
      JSONArray(vectorsJson)
    } catch (e: Exception) {
      throw DmsError.badArgs("selfTest needs a JSON array of vectors")
    }
    var net: GazeNetRunner? = null
    val results = JSONArray()
    for (idx in 0 until vectors.length()) {
      val v = vectors.optJSONObject(idx) ?: JSONObject()
      val name = v.optString("name", "")
      val kind = v.optString("kind", "")
      val inputs = v.optJSONObject("inputs") ?: JSONObject()
      var result = JSONObject().put("name", name).put("kind", kind)
      try {
        when (kind) {
          "record" -> result.put("batch", record(inputs))
          "gazeInputs" -> result.put("frames", gazeInputs(inputs))
          "statsTracker" -> result.put("current", statsTracker(inputs))
          "headPose" -> result.put("poses", headPose(inputs))
          "batcher" -> result.put("cases", batcher(inputs))
          "focal" -> result.put("focalScales", focal(inputs))
          "onnx" -> if (!gazeNetAvailable) {
            result.put("skipped", "gaze net not built")
          } else {
            if (net == null) net = makeNet()
            val runner = net ?: throw DmsError.model("the gaze net could not be created")
            result.put("cases", onnx(inputs, runner))
          }
          else -> throw DmsError.badArgs("unknown vector kind $kind")
        }
      } catch (e: Exception) {
        result = JSONObject().put("name", name).put("kind", kind).put("error", e.message ?: e.toString())
      }
      results.put(result)
    }
    net?.close()
    return JSONObject()
      .put("version", 1)
      .put("platform", PLATFORM)
      .put("gazeNetAvailable", gazeNetAvailable)
      .put("results", results)
      .toString()
  }

  // JSON helpers

  private fun num(o: JSONObject, key: String): Double {
    if (!o.has(key) || o.isNull(key)) throw DmsError.badArgs("expected a number at $key")
    return o.getDouble(key)
  }

  private fun nums(a: JSONArray?): DoubleArray {
    a ?: throw DmsError.badArgs("expected an array of numbers")
    return DoubleArray(a.length()) { a.getDouble(it) }
  }

  /** null → NaN. */
  private fun numsOrNaN(a: JSONArray): DoubleArray = DoubleArray(a.length()) { if (a.isNull(it)) Double.NaN else a.getDouble(it) }

  private fun optionalNums(o: JSONObject, key: String): DoubleArray? =
    if (!o.has(key) || o.isNull(key)) null else nums(o.getJSONArray(key))

  private fun doubles(xs: DoubleArray): JSONArray = JSONArray().also { a -> xs.forEach { a.put(it) } }
  private fun floats(xs: FloatArray): JSONArray = JSONArray().also { a -> xs.forEach { a.put(it.toDouble()) } }

  // Kinds

  private class Image(val w: Int, val h: Int, val stride: Int, val bgra: Boolean, val bytes: ByteBuffer)

  private fun record(inputs: JSONObject): JSONObject {
    val images = inputs.getJSONArray("images")
    val frames = inputs.getJSONArray("frames")
    val anchorEpochMs = num(inputs, "anchorEpochMs")
    val pixels = (0 until images.length()).map { i ->
      val img = images.getJSONObject(i)
      val w = num(img, "w").toInt()
      val h = num(img, "h").toInt()
      val stride = num(img, "stride").toInt()
      val data = B64.decode(img.getString("pixels"))
      if (data.size != stride * h) throw DmsError.badArgs("image bytes ≠ stride·h")
      Image(w, h, stride, img.getString("format") == "bgra", ByteBuffer.wrap(data))
    }
    val records = ArrayList<DoubleArray>(frames.length())
    for (k in 0 until frames.length()) {
      val f = frames.getJSONObject(k)
      val i = num(f, "image").toInt()
      if (i >= pixels.size) throw DmsError.badArgs("record frame")
      val img = pixels[i]
      // The production reader, with the vector's own row stride and a 4-byte pixel (Task 2 review I1).
      val src = LumaSource(img.bytes, img.w, img.h, img.stride, 4, img.bgra)
      records.add(
        FeatureExtractor.buildRecord(
          FrameInput(
            num(f, "tMs"), img.w, img.h, num(f, "rotationDeg").toInt(), src,
            optionalNums(f, "landmarks"), optionalNums(f, "matrix"), optionalNums(f, "netGaze"),
            num(f, "latLandmarkMs"), num(f, "latTotalMs")
          )
        )
      )
    }
    val batch = RecordEncoder.encode(records) ?: throw DmsError.badArgs("record encoding")
    return JSONObject()
      .put("anchorTMs", batch.anchorTMs)
      .put("anchorEpochMs", anchorEpochMs)
      .put("n", batch.n)
      .put("data", B64.encode(batch.data))
  }

  private fun gazeInputs(inputs: JSONObject): JSONArray {
    val width = num(inputs, "width")
    val height = num(inputs, "height")
    val focal = num(inputs, "focalScale")
    val frames = inputs.getJSONArray("frames")
    val asm = GazeInputAssembler()
    val out = JSONArray()
    for (k in 0 until frames.length()) {
      val f = frames.getJSONObject(k)
      val tSec = num(f, "tSec")
      val lm = nums(f.getJSONArray("landmarks"))
      val p = asm.prepare(lm, width, height, focal) ?: throw DmsError.badArgs("degenerate face")
      val g = FeatureExtractor.geometry(lm, width.toInt(), height.toInt())
      val admitted = asm.admit(p.cloud64, tSec, g.right?.ear ?: Double.NaN, g.left?.ear ?: Double.NaN, g.right == null, g.left == null)
      out.put(
        JSONObject()
          .put("cloud", floats(p.cloud))
          .put("context", floats(p.context))
          .put("validity", floats(p.validity))
          .put("admitted", admitted)
      )
    }
    return out
  }

  private fun statsTracker(inputs: JSONObject): JSONArray {
    val mean = nums(inputs.getJSONArray("trainingMean"))
    val t = nums(inputs.getJSONArray("t"))
    val pushes = inputs.getJSONArray("pushes")
    val tracker = SubjectStatisticTracker(mean, num(inputs, "warmup").toInt(), num(inputs, "windowS"))
    val out = JSONArray()
    for (i in 0 until pushes.length()) out.put(doubles(tracker.push(numsOrNaN(pushes.getJSONArray(i)), t[i])))
    return out
  }

  private fun headPose(inputs: JSONObject): JSONArray {
    val cases = inputs.getJSONArray("cases")
    val out = JSONArray()
    for (i in 0 until cases.length()) {
      val c = cases.getJSONObject(i)
      // The production path: the layout rule, then the pose (Task 2 review I2).
      val p = HeadPose.fromAnyLayout(nums(c.getJSONArray("matrix")), num(c, "rotationDeg").toInt())
      out.put(if (p == null) JSONObject.NULL else doubles(p))
    }
    return out
  }

  /** The production Batcher over each case's appends; each flush reports which append triggered it. */
  private fun batcher(inputs: JSONObject): JSONArray {
    val cases = inputs.getJSONArray("cases")
    val out = JSONArray()
    for (i in 0 until cases.length()) {
      val c = cases.getJSONObject(i)
      val interval = num(c, "intervalMs")
      val offset = num(c, "epochOffsetMs")
      val frames = c.getJSONArray("frames")
      val b = Batcher()
      val flushes = JSONArray()
      for (k in 0 until frames.length()) {
        val f = frames.getJSONObject(k)
        val t = num(f, "tMs")
        val now = num(f, "nowMs")
        b.append(FeatureExtractor.faceAbsentRecord(t, 0.0, 0, 0.0, 0.0), now, now + offset)
        if (b.isDue(now, interval)) {
          val n = b.count
          val p = b.flush() ?: throw DmsError.badArgs("batcher flush")
          flushes.put(JSONObject().put("after", k).put("n", n).put("anchorTMs", p["anchorTMs"]).put("anchorEpochMs", p["anchorEpochMs"]))
        }
      }
      out.put(JSONObject().put("flushes", flushes))
    }
    return out
  }

  /** Focal.focalScale per case (sensor fields, or null for the field-of-view fallback). */
  private fun focal(inputs: JSONObject): JSONArray {
    val cases = inputs.getJSONArray("cases")
    val out = JSONArray()
    for (i in 0 until cases.length()) {
      val c = cases.getJSONObject(i)
      val s = if (c.isNull("sensor")) null else c.getJSONObject("sensor").let {
        SensorGeometry(
          num(it, "focalLengthMm"), num(it, "physicalWidthMm"), num(it, "physicalHeightMm"),
          num(it, "pixelArrayWidth").toInt(), num(it, "pixelArrayHeight").toInt(),
          num(it, "activeWidth").toInt(), num(it, "activeHeight").toInt()
        )
      }
      out.put(Focal.focalScale(s, num(c, "width").toInt(), num(c, "height").toInt(), num(c, "rotationDeg").toInt()))
    }
    return out
  }

  private fun onnx(inputs: JSONObject, runner: GazeNetRunner): JSONArray {
    val cases = inputs.getJSONArray("cases")
    val out = JSONArray()
    for (i in 0 until cases.length()) {
      val c = cases.getJSONObject(i)
      fun f32(key: String): FloatArray = nums(c.getJSONArray(key)).let { d -> FloatArray(d.size) { d[it].toFloat() } }
      val (gaze, rotation) = runner.run(f32("cloud"), f32("context"), f32("validity"))
      out.put(JSONObject().put("gaze", doubles(gaze)).put("rotation", doubles(rotation)))
    }
    return out
  }
}

/** Base64 without java.util.Base64 (API 26) or android.util.Base64 (not on a plain JVM). */
internal object B64 {
  private const val ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"

  fun encode(bytes: ByteArray): String {
    val out = StringBuilder((bytes.size + 2) / 3 * 4)
    var i = 0
    while (i < bytes.size) {
      val a = bytes[i].toInt() and 0xFF
      val b = if (i + 1 < bytes.size) bytes[i + 1].toInt() and 0xFF else 0
      val c = if (i + 2 < bytes.size) bytes[i + 2].toInt() and 0xFF else 0
      val n = (a shl 16) or (b shl 8) or c
      out.append(ALPHABET[(n shr 18) and 63]).append(ALPHABET[(n shr 12) and 63])
      out.append(if (i + 1 < bytes.size) ALPHABET[(n shr 6) and 63] else '=')
      out.append(if (i + 2 < bytes.size) ALPHABET[n and 63] else '=')
      i += 3
    }
    return out.toString()
  }

  fun decode(text: String): ByteArray {
    val clean = text.trimEnd('=')
    val out = ByteArray(clean.length * 3 / 4)
    var buffer = 0
    var bits = 0
    var o = 0
    for (ch in clean) {
      val v = ALPHABET.indexOf(ch)
      if (v < 0) throw DmsError.badArgs("invalid base64")
      buffer = (buffer shl 6) or v
      bits += 6
      if (bits >= 8) {
        bits -= 8
        out[o++] = ((buffer shr bits) and 0xFF).toByte()
      }
    }
    return out
  }
}

File: modules/dms-vision/ios/SelfTest.swift
// The native half of the self-test (README §8.3). It runs the PRODUCTION classes (FeatureExtractor,
// Roi, HeadPose, RecordEncoder, Batcher, GazeInputAssembler, SubjectStatisticTracker, GazeNetFactory) over the
// golden vectors and returns their outputs; JS diffs them. It never carries a copy of any production
// logic. Foundation only.

import Foundation

enum SelfTest {
  static let platform = "ios"

  /// `vectorsJson`: a JSON array of vectors. Throws DmsError.badArgs if it is not parseable at all.
  static func run(_ vectorsJson: String) throws -> String {
    guard let data = vectorsJson.data(using: .utf8),
          let vectors = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
      throw DmsError.badArgs("selfTest needs a JSON array of vectors")
    }
    var net: GazeNetRunner? = nil
    let results: [[String: Any]] = vectors.map { v in
      let name = v["name"] as? String ?? ""
      let kind = v["kind"] as? String ?? ""
      let inputs = v["inputs"] as? [String: Any] ?? [:]
      var result: [String: Any] = ["name": name, "kind": kind]
      do {
        switch kind {
        case "record": result["batch"] = try record(inputs)
        case "gazeInputs": result["frames"] = try gazeInputs(inputs)
        case "statsTracker": result["current"] = try statsTracker(inputs)
        case "headPose": result["poses"] = try headPose(inputs)
        case "batcher": result["cases"] = try batcher(inputs)
        case "onnx":
          if !GazeNetFactory.available {
            result["skipped"] = "gaze net not built"
          } else {
            if net == nil { net = try GazeNetFactory.make() }
            guard let runner = net else { throw DmsError.model("the gaze net could not be created") }
            result["cases"] = try onnx(inputs, runner)
          }
        default: throw DmsError.badArgs("unknown vector kind \(kind)")
        }
      } catch let e as DmsError {
        result = ["name": name, "kind": kind, "error": e.message]
      } catch {
        result = ["name": name, "kind": kind, "error": "\(error)"]
      }
      return result
    }
    net?.close()
    let out: [String: Any] = ["version": 1, "platform": platform, "gazeNetAvailable": GazeNetFactory.available, "results": results]
    let bytes = try JSONSerialization.data(withJSONObject: out)
    return String(data: bytes, encoding: .utf8) ?? "{}"
  }

  // MARK: - JSON helpers

  static func num(_ any: Any?) -> Double? {
    if let n = any as? NSNumber { return n.doubleValue }
    if let d = any as? Double { return d }
    if let i = any as? Int { return Double(i) }
    return nil
  }

  static func nums(_ any: Any?) throws -> [Double] {
    guard let arr = any as? [Any] else { throw DmsError.badArgs("expected an array of numbers") }
    return try arr.map { guard let d = num($0) else { throw DmsError.badArgs("expected a number") }; return d }
  }

  /// null → NaN.
  static func numsOrNaN(_ any: Any?) throws -> [Double] {
    guard let arr = any as? [Any] else { throw DmsError.badArgs("expected an array") }
    return arr.map { num($0) ?? .nan }
  }

  static func optionalNums(_ any: Any?) throws -> [Double]? {
    if any == nil || any is NSNull { return nil }
    return try nums(any)
  }

  // MARK: - Kinds

  static func record(_ inputs: [String: Any]) throws -> [String: Any] {
    guard let images = inputs["images"] as? [[String: Any]], let frames = inputs["frames"] as? [[String: Any]],
          let anchorEpochMs = num(inputs["anchorEpochMs"]) else { throw DmsError.badArgs("record inputs") }
    let pixels: [(w: Int, h: Int, stride: Int, bgra: Bool, bytes: [UInt8])] = try images.map { img in
      guard let w = num(img["w"]), let h = num(img["h"]), let stride = num(img["stride"]),
            let format = img["format"] as? String, let b64 = img["pixels"] as? String,
            let data = Data(base64Encoded: b64) else { throw DmsError.badArgs("record image") }
      guard data.count == Int(stride) * Int(h) else { throw DmsError.badArgs("image bytes ≠ stride·h") }
      return (Int(w), Int(h), Int(stride), format == "bgra", [UInt8](data))
    }
    var records: [[Double]] = []
    for f in frames {
      guard let t = num(f["tMs"]), let i = num(f["image"]), Int(i) < pixels.count,
            let rot = num(f["rotationDeg"]), let latL = num(f["latLandmarkMs"]), let latT = num(f["latTotalMs"]) else {
        throw DmsError.badArgs("record frame")
      }
      let img = pixels[Int(i)]
      let landmarks = try optionalNums(f["landmarks"])
      let matrix = try optionalNums(f["matrix"])
      let netGaze = try optionalNums(f["netGaze"])
      let record: [Double] = img.bytes.withUnsafeBufferPointer { buf in
        // The production reader, with the vector's own stride (Task 2 review I1).
        let src = LumaSource(bytes: buf.baseAddress!, width: img.w, height: img.h, rowBytes: img.stride, bgra: img.bgra)
        return FeatureExtractor.buildRecord(FrameInput(
          tMs: t, bufferW: img.w, bufferH: img.h, rotationDeg: Int(rot), luma: src,
          landmarks: landmarks, matrix: matrix, netGaze: netGaze, latLandmarkMs: latL, latTotalMs: latT))
      }
      records.append(record)
    }
    guard let batch = RecordEncoder.encode(records) else { throw DmsError.badArgs("record encoding") }
    return ["anchorTMs": batch.anchorTMs, "anchorEpochMs": anchorEpochMs, "n": batch.n, "data": batch.data.base64EncodedString()]
  }

  static func gazeInputs(_ inputs: [String: Any]) throws -> [[String: Any]] {
    guard let width = num(inputs["width"]), let height = num(inputs["height"]),
          let focal = num(inputs["focalScale"]), let frames = inputs["frames"] as? [[String: Any]] else {
      throw DmsError.badArgs("gazeInputs inputs")
    }
    let asm = GazeInputAssembler()
    return try frames.map { f in
      guard let tSec = num(f["tSec"]) else { throw DmsError.badArgs("gazeInputs frame") }
      let lm = try nums(f["landmarks"])
      guard let prepared = asm.prepare(lm, width, height, focal) else { throw DmsError.badArgs("degenerate face") }
      let g = FeatureExtractor.geometry(lm, Int(width), Int(height))
      let admitted = asm.admit(prepared.cloud64, tSec, earR: g.right?.ear ?? .nan, earL: g.left?.ear ?? .nan,
                               clippedR: g.right == nil, clippedL: g.left == nil)
      return [
        "cloud": prepared.cloud.map { Double($0) },
        "context": prepared.context.map { Double($0) },
        "validity": prepared.validity.map { Double($0) },
        "admitted": admitted,
      ]
    }
  }

  static func statsTracker(_ inputs: [String: Any]) throws -> [[Double]] {
    let mean = try nums(inputs["trainingMean"])
    guard let warmup = num(inputs["warmup"]), let windowS = num(inputs["windowS"]),
          let pushes = inputs["pushes"] as? [Any] else { throw DmsError.badArgs("statsTracker inputs") }
    let t = try nums(inputs["t"])
    let tracker = SubjectStatisticTracker(trainingMean: mean, warmup: Int(warmup), windowS: windowS)
    return try pushes.enumerated().map { i, p in tracker.push(try numsOrNaN(p), t[i]) }
  }

  static func headPose(_ inputs: [String: Any]) throws -> [Any] {
    guard let cases = inputs["cases"] as? [[String: Any]] else { throw DmsError.badArgs("headPose inputs") }
    return try cases.map { c -> Any in
      guard let rot = num(c["rotationDeg"]) else { throw DmsError.badArgs("headPose case") }
      // The production path: the layout rule, then the pose (Task 2 review I2).
      guard let p = HeadPose.fromAnyLayout(try nums(c["matrix"]), Int(rot)) else { return NSNull() }
      return [p.yaw, p.pitch, p.roll]
    }
  }

  /// The production Batcher over each case's appends; each flush reports which append triggered it.
  static func batcher(_ inputs: [String: Any]) throws -> [[String: Any]] {
    guard let cases = inputs["cases"] as? [[String: Any]] else { throw DmsError.badArgs("batcher inputs") }
    return try cases.map { c in
      guard let interval = num(c["intervalMs"]), let offset = num(c["epochOffsetMs"]),
            let frames = c["frames"] as? [[String: Any]] else { throw DmsError.badArgs("batcher case") }
      let b = Batcher()
      var flushes: [[String: Any]] = []
      for (i, f) in frames.enumerated() {
        guard let t = num(f["tMs"]), let now = num(f["nowMs"]) else { throw DmsError.badArgs("batcher frame") }
        b.append(FeatureExtractor.faceAbsentRecord(t, 0, 0, 0, 0), nowMs: now, epochNowMs: now + offset)
        if b.isDue(nowMs: now, intervalMs: interval) {
          let n = b.count
          guard let p = b.flush() else { throw DmsError.badArgs("batcher flush") }
          flushes.append(["after": i, "n": n, "anchorTMs": p["anchorTMs"] ?? NSNull(), "anchorEpochMs": p["anchorEpochMs"] ?? NSNull()])
        }
      }
      return ["flushes": flushes]
    }
  }

  static func onnx(_ inputs: [String: Any], _ runner: GazeNetRunner) throws -> [[String: Any]] {
    guard let cases = inputs["cases"] as? [[String: Any]] else { throw DmsError.badArgs("onnx inputs") }
    return try cases.map { c in
      let out = try runner.run(cloud: try nums(c["cloud"]).map { Float($0) },
                               context: try nums(c["context"]).map { Float($0) },
                               validity: try nums(c["validity"]).map { Float($0) })
      return ["gaze": out.gaze, "rotation": out.rotation]
    }
  }
}
